File: app/src/main/java/com/jajinba/pixabaydemo/adapter/ImageListAdapter.java
package com.jajinba.pixabaydemo.adapter;

import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.jajinba.pixabaydemo.R;
import com.jajinba.pixabaydemo.model.PixabayImageObject;
import com.jajinba.pixabaydemo.utils.ArrayUtils;
import com.jajinba.pixabaydemo.view.ViewHolder;
import com.jajinba.pixabaydemo.view.fragment.ListFragment;

import java.util.ArrayList;
import java.util.List;


public class ImageListAdapter extends RecyclerView.Adapter<ViewHolder> {

  private final int VIEW_TYPE_ITEM = 5566;
  private final int VIEW_TYPE_LOADING = 5567;

  private ListFragment mFragment;
  private List<PixabayImageObject> mImageList;

  public ImageListAdapter(ListFragment fragment) {
    mFragment = fragment;
    mImageList = new ArrayList<>();
  }

  @Override
  public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    return viewType == VIEW_TYPE_ITEM ? new ViewHolder(R.layout.item_image, parent) :
        new ViewHolder(R.layout.item_loading, parent);
  }

  @Override
  public void onBindViewHolder(final ViewHolder holder, int position) {
    if (getItemViewType(position) == VIEW_TYPE_ITEM) {
      if (mFragment.isFragmentValid() && ArrayUtils.getLengthSafe(mImageList) > position) {
        Glide.with(mFragment)
            .load(mImageList.get(position).getWebformatUrl())
            .placeholder(R.drawable.placeholder)
            // workaround, refer glide issue 542 on Github
            .dontAnimate()
            .override(mImageList.get(position).getWebformatWidth(),
                mImageList.get(position).getWebformatHeight())
            .into((ImageView) holder.getView(R.id.image_view));
      }
    } else if (getItemViewType(position) == VIEW_TYPE_LOADING) {
      holder.setVisibility(R.id.load_more_textview, View.VISIBLE);
      holder.setVisibility(R.id.progress_bar, View.GONE);

      // TODO should define setOnClickListener in ViewHolder
      holder.getView(R.id.load_more_textview).setOnClickListener(new View.OnClickListener() {
        @Override
        public void onClick(View view) {
          holder.setVisibility(R.id.load_more_textview, View.GONE);
          holder.setVisibility(R.id.progress_bar, View.VISIBLE);

          mFragment.loadMore();
        }
      });
    }
  }

  @Override
  public int getItemCount() {
    // +1 for bottom loading item
    return ArrayUtils.getLengthSafe(mImageList) + 1;
  }

  @Override
  public int getItemViewType(int position) {
    return ArrayUtils.getLengthSafe(mImageList) > position ? VIEW_TYPE_ITEM : VIEW_TYPE_LOADING;
  }

  public void updateList(List<PixabayImageObject> imageList) {

    Log.e(ImageListAdapter.class.getSimpleName(), "old list: " + ArrayUtils.getLengthSafe(new ArrayList<>(mImageList)));
    Log.e(ImageListAdapter.class.getSimpleName(), "new list: " + ArrayUtils.getLengthSafe(imageList));

    mImageList.clear();
    mImageList.addAll(imageList);
  }

}

File: app/src/main/java/com/jajinba/pixabaydemo/adapter/ImageListAdapter.kt
package com.jajinba.pixabaydemo.adapter

import android.content.Context
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.jajinba.pixabaydemo.Constants
import com.jajinba.pixabaydemo.R
import com.jajinba.pixabaydemo.model.PixabayImageObject
import com.jajinba.pixabaydemo.utils.ArrayUtils
import com.jajinba.pixabaydemo.view.ViewHolder

class ImageListAdapter(private val context: Context?,
                       private val callback: Callback? = null) : RecyclerView.Adapter<ViewHolder?>() {

  companion object {
    private const val VIEW_TYPE_ITEM = 5566
    private const val VIEW_TYPE_LOADING = 5567
  }

  interface Callback {
    fun loadMore()
  }

  private val mImageList = mutableListOf<PixabayImageObject>()

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
    return if (viewType == VIEW_TYPE_ITEM) ViewHolder(
      R.layout.item_image,
      parent
    ) else ViewHolder(R.layout.item_loading, parent)
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    if (getItemViewType(position) == VIEW_TYPE_ITEM) {
      if (ArrayUtils.getLengthSafe(mImageList) > position) {
        context?.let {
          Glide.with(it)
            .load(mImageList[position].webformatUrl)
            .placeholder(R.drawable.placeholder) // workaround, refer glide issue 542 on Github
            .dontAnimate()
            .override(
              mImageList[position].webformatWidth,
              mImageList[position].webformatHeight
            )
            .into(holder.getView<View?>(R.id.image_view) as ImageView)
        }
      }
    } else if (getItemViewType(position) == VIEW_TYPE_LOADING) {
      holder.setVisibility(R.id.load_more_textview, View.VISIBLE)
      holder.setVisibility(R.id.progressBar, View.GONE)

      // TODO should define setOnClickListener in ViewHolder
      holder.getView<View?>(R.id.load_more_textview)?.setOnClickListener {
        holder.setVisibility(R.id.load_more_textview, View.GONE)
        holder.setVisibility(R.id.progressBar, View.VISIBLE)
//        callback?.loadMore()
      }
    }
  }

  override fun getItemCount(): Int {
    // +1 for bottom loading item
    // TODO should check can load more image from API result
    return if (ArrayUtils.getLengthSafe(mImageList) % Constants.IMAGE_PER_PAGE == 0)
      ArrayUtils.getLengthSafe(mImageList) + 1
    else
      ArrayUtils.getLengthSafe(mImageList)
  }

  override fun getItemViewType(position: Int): Int {
    return if (ArrayUtils.getLengthSafe(mImageList) > position) VIEW_TYPE_ITEM else VIEW_TYPE_LOADING
  }

  fun updateList(imageList: MutableList<PixabayImageObject>) {
    Log.d(ImageListAdapter::class.java.simpleName, "updateList with ${imageList.size} images")
    mImageList.clear()
    mImageList.addAll(imageList)
    Log.d(ImageListAdapter::class.java.simpleName, "${mImageList.size} images after update")
  }
}
